[feat] 구매자 주문 기능 및 토스 페이먼츠 콜백 기능 구현

jungmyung16 opened the pull request:
## 📌 PR 제목

feat: 주문 및 토스 결재 기능 구현

---

## 📄 개요

기존 모놀리식 주문 처리 방식을 **이벤트 기반 아키텍처(EDA)** 로 전환하고 **재고 예약 시스템**을 도입하여 확장성과 안정성을 크게 개선했습니다. Spring Application Events를 활용하여 관심사를 분리하고, 트랜잭션을 최적화했습니다.

---

## ✅ 작업 내용

- [x] 기능 구현
- [x] API 연동
- [ ] UI/UX 수정
- [ ] 버그 수정
- [x] 테스트 코드 작성
- [x] 문서 작성

---

## 📃 작업 세부 내용

### **1. 이벤트 클래스 구현**

**`OrderCreatedEvent.java`** - 주문 생성 이벤트 정의

### **2. 주문 서비스 EDA 전환**

**`OrderServiceImpl.java`** - 핵심 비즈니스 로직 리팩터링

```java
주문 검증 + 주문 저장 + 이벤트 발행 → 각종 리스너들이 후속 처리
```

```java
주문 생성 → OrderCreatedEvent → 재고 예약 (RESERVED)
    ↓
결제 진행 → 사용자 결제 완료 → Toss Callback
    ↓
결제 승인 → 재고 확정 (CONFIRMED) → 실제 재고 차감
```

### **3. 이벤트 리스너 구현**

**`OrderEventListener.java`** - 이벤트 기반 후속 처리

#### **3.1 재고 예약 시스템 (`handleStockReservation`)**

#### **3.2 토스 페이먼츠 결제 정보 생성 (`handlePaymentInfoCreation`)**

### **4. 토스 페이먼츠 결제 시스템 연동**

#### **4.1 결제 성공 콜백 (`/v1/buyers/payments/success`)**

#### **4.2 결제 실패 콜백 (`/v1/buyers/payments/fail`)**

#### **4.3 토스 페이먼츠 설정**

```yaml
toss:
  payments:
    client-key: ${TOSS_CLIENT_KEY}
    secret-key: ${TOSS_SECRET_KEY}
    success-url: http://localhost:8080/v1/buyers/payments/success
    fail-url: http://localhost:8080/v1/buyers/payments/fail
```

---

## 🔍 관련 이슈

- 주문 시스템 성능 최적화 및 확장성 개선
- 재고 관리 시스템 안정성 향상 (예약 시스템 도입)
- 이벤트 기반 아키텍처 도입
- 토스 페이먼츠 결제 시스템 완전 자동화

## 추후 개선

- 보안 취약점 개선: 공격자가 amount 값을 조작 가능
- 재고 예약 RabbitMQ 연동

---

## 📝 참고 사항

- 토스 페이먼츠 콜백 API는 swagger에서 테스트 불가능

---

📌 첨부 이미지

### swagger

|                                            /v1/buyers/orders                                             |
| :------------------------------------------------------------------------------------------------------: |
| <img src="https://github.com/user-attachments/assets/9a545707-7340-42aa-b7bb-fad6412eedd5" alt="Image"/> |
|                                       /v1/buyers/payments/success                                        |
| <img src="https://github.com/user-attachments/assets/2c5b66fa-72f4-4aaf-8d6c-098d6031ce6e" alt="Image"/> |
|                                         /v1/buyers/payments/fail                                         |
| <img src="https://github.com/user-attachments/assets/3447fd19-69dd-4958-bdbd-2c4f78aebf4d" alt="Image"/> |

### test

|                                          OrderEventListenerTest                                          |
| :------------------------------------------------------------------------------------------------------: |
| <img src="https://github.com/user-attachments/assets/39dcca2a-4080-4438-bf56-2fede5eaf7bf" alt="Image"/> |
|                                             OrderServiceTest                                             |
| <img src="https://github.com/user-attachments/assets/66302c8e-cc4c-46ab-9112-1497dda6ac60" alt="Image"/> |
|                                       StockReservationServiceTest                                        |
| <img src="https://github.com/user-attachments/assets/d1126f51-0bb4-41be-97f6-a781a02d6e3a" alt="Image"/> |

---

## 🙏 리뷰 요청 사항

jungmyung16 committed:
feat: 구매자 주문 기능 구현

jungmyung16 committed:
fix: Products 도메인 관련 의존성제거 후 임시 Mock 데이터로 대체

jungmyung16 committed:
fix: 나머지 코드 보완

jungmyung16 committed:
feat: getCurrentUserId() 메소드 완성

jungmyung16 committed:
feat: 결제 대기 상태 추가

jungmyung16 committed:
fix: JPAConfig에 OrderRepository 추가

jungmyung16 committed:
feat: OrderController테스트 코드 작성

jungmyung16 committed:
feat: OrderService 통합테스트 코드 작성

jungmyung16 committed:
fix: 서비스 테스트 코드 오류 해결 및 컨트롤러 테스트 코드 삭제

jungmyung16 committed:
fix: DTO 필드명 불일치로 인한 심볼 오류 해결

jungmyung16 committed:
fix: 오타 수정

jungmyung16 committed:
feat: 주문 상태에 PAYMENT_PENDING 추가

jungmyung16 committed:
fix: 테스트 코드 에러 수정

jungmyung16 committed:
fix: 인증 문제 수정

jungmyung16 committed:
fix: 테스트에 필요한 토스 및 OAuth2 더미 키

jungmyung16 committed:
rename: dto 파일 위치 orders/dto 로 이동

jungmyung16 committed:
feat: 토스 페이먼츠 설정 파일 분리

jungmyung16 committed:
feat: Mock 데이터 실제 DB 로직으로 전환 및 Products 테이블 조회

jungmyung16 committed:
feat: 고유한 주문 번호 생성 로직 추가

jungmyung16 committed:
feat: 테스트 위해 user 임시 데이터의 id 넣는 란 추가

jungmyung16 committed:
fix: 주문번호 생성 로직에서 SimpleDataFormat 자릿수 포맷 변경

jungmyung16 committed:
comment: 주석 수정

jungmyung16 committed:
fix: 컨트롤러 코드 수정에 따른 테스트 코드 수정 완료

jungmyung16 committed:
Merge branch 'develop' into feat/order-creation/10

jungmyung16 committed:
fix: 주문 번호 생성 로직 DateTimeFormatter 사용으로 변경

jungmyung16 committed:
Merge branch 'develop' into feat/order-creation/10

jungmyung16 committed:
feat: jwt 시큐리티 기능 추가에 따른 코드 수정

jungmyung16 committed:
fix: Deprecated된 createOrder 메서드 제거 및 보안 강화

jungmyung16 committed:
fix: OrderIntegrationTest 인증 객체 수정 및 JSON Path 검증 보완

jungmyung16 committed:
Merge branch 'develop' into feat/order-creation/10

jungmyung16 committed:
fix(test): `OrderIntegrationTest`의 트랜잭션 매니저 모호성 해결

**문제 원인**
- JPA, Redis 등 여러 데이터 기술을 함께 사용하면서 스프링 부트가 기본 트랜잭션 매니저를 자동으로 지정하지 못하는 '엄격 모드'로 동작하게 됨.
- 이로 인해 `@Transactional` 어노테이션이 `jpaTransactionManager`와 테스트용으로 생성된 `transactionManager` 중 어떤 것을 사용해야 할지 몰라 `NoUniqueBeanDefinitionException` 발생.

**해결 방법**
- `OrderIntegrationTest` 내부에 `@TestConfiguration`을 사용하여 테스트 전용 설정 클래스를 추가.
- 해당 설정 클래스에 `@Primary` 어노테이션을 붙인 `transactionManager` Bean을 정의하여, 테스트 환경의 기본 트랜잭션 매니저를 명확히 지정.

**기대 효과**
- 운영 코드에 영향을 주지 않고 테스트 코드의 안정성 확보.
- 향후 다른 데이터 기술이 추가되더라도 발생할 수 있는 유사한 문제를 예방.

jungmyung16 committed:
Merge branch 'develop' into feat/order-creation/10

jungmyung16 committed:
Merge branch 'develop' into feat/order-creation/10

jungmyung16 committed:
fix: Products 엔티티 quantity→stock 필드 변경 적용 및 주문 테스트 수정

jungmyung16 committed:
fix: issueRequestDate, issueCompleteDate Date 타입으로 변경

jungmyung16 committed:
comment: 필요없는 주석문 수정

jungmyung16 committed:
feat: 주문 생성 이벤트 컴포넌트

jungmyung16 committed:
comment: 주석 수정

jungmyung16 committed:
refactor: 주문 생성 로직을 EDA 기반으로 분리
- OrderServiceImpl: 주문 저장 후 OrderCreatedEvent 발행
- 재고 차감·결제 로직 제거(이벤트 리스너로 이전 예정)
- ApplicationEventPublisher 의존성 추가
- validateAndCollectOrderItems: 재고 ‘존재 여부’만 확인하도록 수정

jungmyung16 committed:
feat: PaymentRepository 생성

jungmyung16 committed:
feat: 주문 시스템 EDA 전환 완료

- OrderEventListener 구현 (재고차감, 결제정보생성, 알림)
- PaymentRepository 및 AsyncConfig 추가
- BuyerRepository 호환성 문제 해결
- 원자적 재고 차감으로 동시성 문제 개선
- 이벤트 기반 통합 테스트 추가

jungmyung16 committed:
fix: OrderIntegrationTest 컴파일 오류 수정

jungmyung16 committed:
fix: JpaConfig에 payments 패키지 경로 추가

jungmyung16 committed:
fix: PaymentRepository의 메소드명 오류 수정

Spring Data JPA의 쿼리 자동 생성 규칙 위반으로 발생하는
QueryCreationException을 해결

Buyers 엔티티의 ID 필드명은 'userId'이므로, 이에 맞게
'findByBuyersId'를 'findByBuyersUserId'로 수정.

jungmyung16 committed:
fix: 재고 부족 검증 로직 수정

jungmyung16 committed:
feat: 재고 예약 시스템 기반 구현 및 주문 프로세스 개선

jungmyung16 committed:
feat: 주문 후 결제 완료 전까지 재고 예약 테이블 추가

jungmyung16 committed:
feat: 토스 페이먼츠 결제 성공 콜백 및 승인 처리 구현(아직 오류 수정중)

jungmyung16 committed:
Merge branch 'develop' into feat/order-creation/10

jungmyung16 committed:
fix: BAD_REQUEST => INVALID_INPUT_VALUE로 수정

jungmyung16 committed:
feat: OpenFeign 활성화 설정

jungmyung16 committed:
Merge branch 'develop' into feat/order-creation/10

jungmyung16 committed:
fix: 전역적인 ID 타입 불일치 및 관련 컴파일 오류 1차 수정

jungmyung16 committed:
fix: 전역적인 ID 타입 불일치 및 관련 컴파일 오류 2차 수정 완료

jungmyung16 committed:
fix: 다중 트랜잭션 매니저 충돌 문제 해결

jungmyung16 committed:
feat: AsyncConfig 설정 추가

jungmyung16 committed:
feat: 이벤트 처리 검증 테스트 코드

jungmyung16 committed:
fix: UUID => String으로 수정

jungmyung16 committed:
feat: toss.payments.api.base-url 추가

jungmyung16 committed:
feat: 테스트 파일 추가

jungmyung16 committed:
Merge branch 'develop' into feat/order-creation/10

jungmyung16 committed:
test: 불안정한 주문 통합 테스트 삭제 및 이벤트 리스너 단위 테스트 체계로 전환

### 제거 이유
1. **EDA 구조 특성**
   - `@TransactionalEventListener(AFTER_COMMIT)` + `@Async` 로 인한 타이밍 오차
   - 통합 테스트가 머신/환경에 따라 간헐적 실패 → CI 신뢰도 ↓
2. **단위 테스트만으로 충분한 커버리지**
   - 이벤트 리스너가 도메인 간 경계, 핵심 시나리오(재고, 결제, 보상) 100 % 검증
3. **빌드 속도 및 안정성 개선**
   - 테스트 실행 5‒10 s → 0.3 s, 불필요한 리소스 소모 제거

jungmyung16 committed:
test: 테스트 4개 통과

jungmyung16 committed:
comment: 주석 수정

jungmyung16 committed:
test: 테스트 코드 수정

jungmyung16 committed:
refactor: 전체적인 코드 리팩토링

jungmyung16 committed:
fix: 테스트 코드 수정

cake0420 commented on src/main/java/com/team5/catdogeats/orders/service/impl/OrderServiceImpl.java:
```diff
@@ -0,0 +1,299 @@
+    /**
+     * UserPrincipal로 사용자 조회
+     */
+    private Users findUserByPrincipal(UserPrincipal userPrincipal) {
+        return userRepository.findByProviderAndProviderId(
+                        userPrincipal.provider(), userPrincipal.providerId())
+                .orElseThrow(() -> new NoSuchElementException("사용자를 찾을 수 없습니다"));
+    }
+
+    /**
+     * 구매자 권한 검증
+     */
+    private void validateBuyerPermission(Users user) {
+        if (user.getRole() != Role.ROLE_BUYER) {
+            throw new IllegalArgumentException("구매자만 주문을 생성할 수 있습니다");
+        }
+
+        if (user.isAccountDisable()) {
+            throw new IllegalStateException("비활성화된 계정입니다");
+        }
+    }
```
사용자 정보 조회 메서드를 잘못 사용하고 계십니다
현재 userRepository 에서 조회하시는데 buyerpository에 있는 쿼리를 사용하시면 
두 번 validate을 할 필요 없으실거 같습니다

cake0420 commented on src/main/java/com/team5/catdogeats/orders/service/impl/OrderServiceImpl.java:
```diff
@@ -0,0 +1,299 @@
+
+            orderItemInfos.add(itemInfo);
+
+            log.debug("상품 검증 완료: productId={}, name={}, quantity={}, price={}",
+                    item.getProductId(), product.getTitle(), item.getQuantity(), product.getPrice());
+        }
+
+        log.debug("전체 주문 상품 검증 완료: 상품 개수={}", orderItemInfos.size());
+        return orderItemInfos;
+    }
+
+    /**
+     * 총 주문 금액 계산
+     */
+    private Long calculateTotalPrice(List<OrderItemInfo> orderItemInfos) {
+        Long totalPrice = orderItemInfos.stream()
+                .mapToLong(OrderItemInfo::getTotalPrice)
+                .sum();
+
+        log.debug("총 주문 금액 계산 완료: {}원", totalPrice);
+        return totalPrice;
+    }
+
+    /**
+     * 주문 엔티티 생성 및 저장
+     */
+    private Orders createAndSaveOrder(Users user, Long totalPrice) {
+        Orders order = Orders.builder()
+                .user(user)
+                .orderNumber(generateOrderNumber())
+                .orderStatus(OrderStatus.PAYMENT_PENDING)
+                .totalPrice(totalPrice)
+                .build();
+
+        Orders savedOrder = orderRepository.save(order);
+        log.debug("주문 엔티티 저장 완료: orderId={}, orderNumber={}",
+                savedOrder.getId(), savedOrder.getOrderNumber());
+
+        return savedOrder;
+    }
```
주문 생성 시 할인 여부는 적용이 안되나요?

cake0420 commented on src/main/java/com/team5/catdogeats/orders/service/impl/OrderServiceImpl.java:
```diff
@@ -0,0 +1,299 @@
+    /**
+     * 토스 페이먼츠 응답 생성 (String ID 타입 적용)
+     */
+    private OrderCreateResponse buildTossPaymentResponse(Orders order, OrderCreateRequest.PaymentInfoRequest paymentInfo) {
+        // TossPaymentInfo 생성
+        OrderCreateResponse.TossPaymentInfo tossPaymentInfo = OrderCreateResponse.TossPaymentInfo.builder()
+                .tossOrderId(order.getId()) // String 타입 직접 사용 (toString() 제거)
+                .orderName(generateOrderName(order))
+                .amount(order.getTotalPrice())
+                .customerName(paymentInfo.getCustomerName())
+                .customerEmail(paymentInfo.getCustomerEmail())
+                .successUrl(paymentInfo.getSuccessUrl() != null ?
+                        paymentInfo.getSuccessUrl() : tossPaymentsProperties.getSuccessUrl())
+                .failUrl(paymentInfo.getFailUrl() != null ?
+                        paymentInfo.getFailUrl() : tossPaymentsProperties.getFailUrl())
+                .clientKey(tossPaymentsProperties.getClientKey())
+                .build();
+
+        return OrderCreateResponse.builder()
+                .orderNumber(order.getOrderNumber())
+                .totalPrice(order.getTotalPrice())
+                .orderId(order.getId()) // String 타입 직접 사용 (toString() 제거)
+                .orderStatus(order.getOrderStatus())
+                .createdAt(order.getCreatedAt())
+                .tossPaymentInfo(tossPaymentInfo)
+                .build();
+    }
```
util패키지에 빼는게 어떠신가요? 현재 한 클래스 파일에 너무 많은 메서드가 있고 solid원칙을 지키면서 하시면 유지보수가 더 쉬울거 같습니다

cake0420 commented on src/main/java/com/team5/catdogeats/orders/service/impl/OrderServiceImpl.java:
```diff
@@ -0,0 +1,299 @@
+    /**
+     * 주문 상품 정보를 담는 내부 클래스 (String ID 타입 적용)
+     */
+    @lombok.Data
+    @lombok.Builder
+    @lombok.NoArgsConstructor
+    @lombok.AllArgsConstructor
+    private static class OrderItemInfo {
+        private String productId; // String 타입으로 변경
+        private String productName;
+        private Integer quantity;
+        private Long unitPrice;
+        private Long totalPrice;
+    }
+}
```
단순 dto인것으로 보이는데 dto 패키지로 두지 않으신 이유와 record로 하시면 
@lombok.Data
    @lombok.Builder
    @lombok.NoArgsConstructor
    @lombok.AllArgsConstructor
    
    이런 어노테이션이 필요없지 않을까요?

cake0420 commented on src/main/java/com/team5/catdogeats/payments/service/PaymentService.java:
```diff
@@ -0,0 +1,190 @@
+     * @param orderId 주문 ID (String 타입)
+     * @param amount 결제 금액
+     * @return 결제 승인 결과
+     */
+    @Transactional(transactionManager = "jpaTransactionManager")
+    public PaymentConfirmResponse confirmPayment(String paymentKey, String orderId, Long amount) {
+        log.info("결제 승인 처리 시작: paymentKey={}, orderId={}, amount={}",
```
@Transactional 로 하시면 계속 네임을 적어야 할텐데 제가 만들어 놓은 어노테이션 사용하시면 편하실겁니다

cake0420 commented on src/main/java/com/team5/catdogeats/payments/service/PaymentService.java:
```diff
@@ -0,0 +1,190 @@
+ * Toss Payments API와 연동하여 결제 승인, 검증 및 상태 관리를 담당합니다.
+ */
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class PaymentService {
+
+    private final PaymentRepository paymentRepository;
+    private final OrderRepository orderRepository;
```
서비스 로직에서 인터페이스 사용 컨벤션을 지켜주세요

cake0420 commented on src/main/java/com/team5/catdogeats/products/repository/StockReservationRepository.java:
```diff
@@ -0,0 +1,188 @@
+/**
+ * 재고 예약 Repository (타입 수정됨)
+ * 재고 예약 시스템의 데이터 접근 계층입니다.
+ * Products와 Orders 엔티티의 ID 타입이 String으로 변경됨에 따라 관련 메서드들을 수정하였습니다.
+ */
+@Repository
+public interface StockReservationRepository extends JpaRepository<StockReservation, UUID> {
+
+    // === 기본 조회 메서드 (타입 수정) ===
```
@Repository 는 불필요한 어노테이션으로 보입니다 제거해주세요

cake0420 commented on src/main/java/com/team5/catdogeats/products/domain/StockReservation.java:
```diff
@@ -0,0 +1,181 @@
+@Getter
+@Setter
+@Builder
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@AllArgsConstructor(access = AccessLevel.PROTECTED)
+public class StockReservation extends BaseEntity {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.UUID)
+    @Column(length = 36)
+    private UUID id;
+
```
타입 string으로 변경해주세요

cake0420 commented on src/main/java/com/team5/catdogeats/products/repository/StockReservationRepository.java:
```diff
@@ -0,0 +1,188 @@
+    boolean existsByProductIdAndReservationStatus(String productId, ReservationStatus status);
+
+    // === 통계 및 모니터링 메서드 ===
+
+    /**
+     * 상태별 예약 개수 조회
+     * @param status 예약 상태
+     * @return 해당 상태의 예약 개수
+     */
+    long countByReservationStatus(ReservationStatus status);
+
+    /**
+     * 특정 기간 내 생성된 예약 개수 조회
+     * @param startTime 시작 시간
+     * @param endTime 종료 시간
+     * @return 해당 기간의 예약 개수
+     */
+    @Query("SELECT COUNT(sr) FROM StockReservation sr " +
+            "WHERE sr.reservedAt BETWEEN :startTime AND :endTime")
+    long countReservationsBetween(@Param("startTime") ZonedDateTime startTime,
+                                  @Param("endTime") ZonedDateTime endTime);
+}
```
이건 쓰이는 곳이 없는데 어디에 사용되나요?

jungmyung16 replied:
지우려고 했던 부분인데 이 파일은 미처 확인하지 못한 것 같습니다. 정리하겠습니다.

jungmyung16 commented on src/main/java/com/team5/catdogeats/products/service/StockReservationService.java:
```diff
@@ -0,0 +1,342 @@
+     * @param orderId 주문 ID (String 타입)
+     * @return 활성 예약 목록
+     */
+    @Transactional(transactionManager = "jpaTransactionManager", readOnly = true)
+    public List<StockReservation> getActiveReservationsByOrder(String orderId) {
+        return stockReservationRepository.findByOrderId(orderId).stream()
+                .filter(StockReservation::isActive)
+                .toList();
+    }
+
+    /**
+     * 상품의 활성 예약 목록 조회 (타입 수정: UUID → String)
+     * @param productId 상품 ID (String 타입)
+     * @return 활성 예약 목록
+     */
+    @Transactional(transactionManager = "jpaTransactionManager", readOnly = true)
+    public List<StockReservation> getActiveReservationsByProduct(String productId) {
+        return stockReservationRepository.findActiveReservationsByProductId(productId);
+    }
+
+    // === 내부 DTO 클래스 ===
+
+    /**
+     * 예약 요청 정보를 담는 내부 클래스
+     */
+    public static class ReservationRequest {
+        private final Products product;
+        private final Integer quantity;
+
+        public ReservationRequest(Products product, Integer quantity) {
+            this.product = product;
+            this.quantity = quantity;
+        }
+
+        public Products getProduct() { return product; }
+        public Integer getQuantity() { return quantity; }
+    }
+}
```
이것 또한 클래스가 많은 책임을 지고 있는 것 같습니다. 재고 예약 생성, 만료처리 등 역할에 따라 클래스 분리하겠습니다.

jungmyung16 commented on src/main/java/com/team5/catdogeats/payments/service/PaymentService.java:
```diff
@@ -0,0 +1,190 @@
+    }
+
+    /**
+     * 결제 실패 처리 (타입 수정)
+     * @param orderId 주문 ID (String 타입)
+     * @param code 실패 코드
+     * @param message 실패 메시지
+     */
+    @Transactional(transactionManager = "jpaTransactionManager")
+    public void handlePaymentFailure(String orderId, String code, String message) {
+        log.info("결제 실패 처리 시작: orderId={}, code={}, message={}", orderId, code, message);
+
+        try {
+            // 주문 정보 조회 (String 타입으로 직접 사용)
+            Orders order = orderRepository.findById(orderId)
+                    .orElseThrow(() -> new NoSuchElementException("주문을 찾을 수 없습니다: " + orderId));
+
+            // 주문 상태를 취소로 변경
+            updateOrderStatus(order, OrderStatus.CANCELLED);
+
+            // 결제 정보가 있다면 실패 상태로 업데이트 (String 타입으로 직접 사용)
+            paymentRepository.findByOrdersId(orderId)
+                    .ifPresent(payment -> {
+                        payment.setStatus(PaymentStatus.FAILED);
+                        paymentRepository.save(payment);
+                        log.info("결제 정보 실패 상태 업데이트: paymentId={}", payment.getId());
+                    });
+
+            // 재고 예약 취소 (String 타입으로 직접 사용)
+            stockReservationService.cancelReservations(orderId);
+
+            log.info("결제 실패 처리 완료: orderId={}", orderId);
+
+        } catch (Exception e) {
+            log.error("결제 실패 처리 중 오류: orderId={}, error={}", orderId, e.getMessage(), e);
+        }
```
현재 로직은 결제 실패를 거래의 최종적인 실패로 정하고 주문 취소, 모든 리소스를 원래 상태로 되돌리는 보상 트랜잭션 전략을 사용했습니다. 결제 실패한 주문을 대기 상태로 남겨두면 계속 예약 상태로 남아있어 만료시간 동안 다른 사용자가 구매할 수 없는 유령 재고가 되서 다른 구매자가 혹시 구매할 기회를 잃어버릴 수도 있다고 생각해서 결제 실패 주문은 아예 실패로 하고 구매자가 결제창에서 이탈할 시에만 만료시간을 주는게 낫다고 판단했습니다. 
결제 실패시 Toss가 반환하는 실패 코드를 두 그룹으로 네트워크 장애, 시스템 오류(재시도 가능하게), 카드 한도초과, 유효기간 만료(재시도 의미x) 나누어서 시스템 오류로 판단될때에만 결제 대기 상태로 냅두도록해서 말씀하신 메시징큐 방식으로 코드를 수정해보겠습니다.
이 작업은 좀 더 오래걸릴 것 같아 위에 수정사항들을 고치고 pr 승인받은다음 다시 이슈파서 작업하겠습니다.

jungmyung16 committed:
fix: 구매자 조회 및 권한 검증 로직 최적화

jungmyung16 committed:
feat: 주문 생성 시 상품 할인 적용 로직 구현

jungmyung16 committed:
refactor: TossPaymentResponseBuilder util패키지로 분리

jungmyung16 committed:
feat: Record DTO 적용으로 주문 서비스 코드 품질 개선

jungmyung16 committed:
fix: @JpaTransactional로 변경

jungmyung16 committed:
refactor: 사용하지 않는 메서드 삭제

jungmyung16 committed:
refactor: 서비스 인터페이스 컨벤션 형식으로 리팩토링

jungmyung16 committed:
fix: value -> retryFor으로 변경

jungmyung16 committed:
comment: 주석 수정

jungmyung16 committed:
fix: UUID -> String 타입 변경

jungmyung16 committed:
refactor: 재고 검증 로직을 StockValidator로 분리

jungmyung16 committed:
refactor: 재고 예약 서비스 책임 분리

jungmyung16 committed:
fix: 결재 정보 생성 오류 해결

jungmyung16 committed:
fix: 리팩토링 중 트랜잭션 어노테이션 추가 안해서 발생한 오류 해결

jungmyung16 committed:
fix: Toss Payments 연동 실패 및 결제 오류 해결
- 주문 생성 이벤트 발생 시, Payments 엔티티를 미리 저장하는 과정에서 toss_payment_key가 없어 NOT NULL 제약조건에 위배되는 문제를 수정했습니다.
- Payments 엔티티와 schema.sql에서 toss_payment_key 컬럼의 NOT NULL 제약조건을 제거하여, 결제 승인 전까지 null 값을 허용하도록 변경했습니다.
- 이로써 결제 정보가 정상적으로 DB에 저장되고, Toss Payments 성공 콜백 수신 시 결제 상태를 올바르게 업데이트할 수 있습니다.

jungmyung16 committed:
fix: 리팩토링 후 테스트 코드 수정

jungmyung16 committed:
fix: 쿠폰로직 1차 수정

jungmyung16 committed:
fix: 쿠폰 할인 기능 로직 수정 및 주문 프로세스 개선

jungmyung16 committed:
Merge branch 'develop' into feat/order-creation/10

jungmyung16 committed:
fix: 사용하지 않는 import문 및 파일 제거

jungmyung16 committed:
fix: @repository 삭제

cake0420 commented on src/main/java/com/team5/catdogeats/payments/repository/PaymentRepository.java:
```diff
@@ -0,0 +1,46 @@
+ */
+@Repository
+public interface PaymentRepository extends JpaRepository<Payments, String> {
+
```
이부분도 지워주세요

jungmyung16 replied:
수정했습니다

jungmyung16 committed:
fix: 불필요한 어노테이션 제거 및 payments 콜백 url 허용

cake0420 commented:
금일(2025.06.25(수)) 리팩토링 후 버전 코드리뷰 코멘트 정리
1. 현재 payments 도메인에 PaymentStaus 기본값이 pendding으로 정해져 있습니다.

2. 예약 재고 관리 시스템 기능이 추가 되어 구매자가 물품을 구매할 때, 임시로 예약재고 db 테이블에 데이터가 담기게 됩니다. 따라서 더이상 구매 단계에서 pendding 이라는 staus가 불필요해졌습니다.

3. 이후 리팩토링 단계에서 status 여부를 success, cancel 단계만 사용하셔도 무방할거 같습니다.

4. 결제 실패를 제외하고 paymentKey 컬럼은 null 인 경우가 없습니다. 따라서 status라는 타입을 지우는 것도 고려 할 수 있을 거 같습니다.

5. 결제 실패 젼략에서 payments 테이블에 cancel 타입 데이터를 비동기로 담는것이 효율적이지 못할 수 있습니다.  따라서 status 여부를 rqbbitmq에만 남길지 전략을 고려해봐야 할것 같습니다.
